Add java sol to usaco superbull

TruaShamu opened the pull request.

SansPapyrus683 commented on solutions/usaco-531.mdx:
````diff
@@ -121,4 +121,103 @@ int main()
+
+```cpp
+#include <iostream>
+#include <cstdio>
````
wait why this

maggie-j-liu replied:
freopen

SansPapyrus683 commented on solutions/usaco-531.mdx:
```diff
@@ -121,4 +121,103 @@ int main()
+	});
+	int eliminated = 0;
+	long long ans = 0;
+	for (Edge e : edges)
```
maybe add a short comment "Run Kruskal's on the edges" or something like that idk

SansPapyrus683 commented on solutions/usaco-531.mdx:
```diff
@@ -121,4 +121,103 @@ int main()
+		return a.weight > b.weight;
+	});
+	int eliminated = 0;
+	long long ans = 0;
```
(not necessary) `max_score` or smth? maybe?

SansPapyrus683 reviewed (changes requested):
~~also put the dsu stuff in a class lmao~~

maggie-j-liu commented on solutions/usaco-531.mdx:
```diff
@@ -121,4 +121,103 @@ int main()
 
 </CPPSection>
 
+<CPPSection>
```
looks like you put the wrong code? (this is mine 👀 )

nathangong commented:
Is this java code? It seems like youre just adding another C++ code block

TruaShamu commented:
Sorry, I was dumb and I put the wrong code.
It should be fixed now.

maggie-j-liu commented on solutions/usaco-531.mdx:
```diff
@@ -121,4 +121,59 @@ int main()
 
 </CPPSection>
 
+<JavaSection>
```
this is prim's and the c++ sol is kruskal's, so put this in a separate language section (and maybe add headers to the sections)

maggie-j-liu commented on solutions/usaco-531.mdx:
````diff
@@ -121,4 +121,59 @@ int main()
+<JavaSection>
+
+```java
+import java.io.*;
````
don't you need to `import java.util.*` ?

maggie-j-liu commented on solutions/usaco-531.mdx:
```diff
@@ -121,4 +121,59 @@ int main()
+	public static void main(String[] args) throws IOException {
+		Kattio io = new Kattio("superbull");
+		int N = io.nextInt();
+		//Read in the input array.
```
```suggestion
		// Read in the input array.
```

SansPapyrus683 commented on solutions/usaco-531.mdx:
````diff
@@ -121,4 +121,59 @@ int main()
+```java
+import java.io.*;
+
+public class superbull {
````
```suggestion
public class Superbull {
```

SansPapyrus683 commented on solutions/usaco-531.mdx:
```diff
@@ -121,4 +121,59 @@ int main()
+		Kattio io = new Kattio("superbull");
+		int N = io.nextInt();
+		//Read in the input array.
+		int[] inp = new int[N];
```
better name for `inp` maybe?

SansPapyrus683 commented on solutions/usaco-531.mdx:
```diff
@@ -121,4 +121,59 @@ int main()
+			inp[i] = io.nextInt();
+		}
+
+		long ans = 0;
```
(not necessary) maybe `max_cost` instead?

SansPapyrus683 commented on solutions/usaco-531.mdx:
```diff
@@ -121,4 +121,59 @@ int main()
+				score[i] = Integer.max(score[i], (inp[team2] ^ inp[i]));
+			}
+		}
+
```
```suggestion
```

TruaShamu commented:
Alright, done

TruaShamu commented:
Ok, I also renamed ans -> max_cost and renamed inp -> id.

maggie-j-liu commented on solutions/usaco-531.mdx:
```diff
@@ -122,3 +122,64 @@ int main()
 </LanguageSection>
+
+
+## Prim's Implementation
```
add time complexity

SansPapyrus683 reviewed (approved):
nice
i also just realized where your username came from

ryanchou-dev commented on solutions/usaco-531.mdx:
```diff
@@ -2,7 +2,7 @@
+author: Maggie Liu, Sofia Yang
 ---
 
 [Official Analysis](http://www.usaco.org/current/data/sol_superbull_silver.html)
```
```suggestion
[Official Analysis (C++)](http://www.usaco.org/current/data/sol_superbull_silver.html)
```

ryanchou-dev reviewed (changes requested):
just this :)

dongliu0426 commented on solutions/usaco-531.mdx:
```diff
@@ -122,3 +122,64 @@ int main()
+				score[i] = Integer.max(score[i], (ids[team2] ^ ids[i]));
+			}
+		}
+		pw.println(max_cost);
+		pw.close();
```
io?

dongliu0426 replied:
was this tested?

TruaShamu replied:
fixed

TruaShamu replied:
it works
[image](https://ibb.co/wgLHRdn)
[code](https://pastebin.com/fErXW15Y)

dongliu0426 commented on solutions/usaco-531.mdx:
```diff
@@ -122,3 +122,66 @@ int main()
+		boolean[] visited = new boolean[N];
+		int[] score = new int[N];
+
+		// Loop through possible choices for the first team.
+		for (int team1 = 0; team1 < N; team1++) {
```
:/ I don't think prim
> Loop through possible choices for the first team.

Doesn't it loop through N iterations, keeping track of node with the minimum (or maximum in this case) weight? It doesn't have anything to do with `team1`?

TruaShamu commented:
I've changed the variable names, rewrote the comments and also added an explanation to make it seem more intuitive. I hope you're happy and satisfied 😄 

I think it's Prim. If you still think it isn't, what do you think it should be, then?

maggie-j-liu commented:
> I think it's Prim. If you still think it isn't, what do you think it should be, then?

it is prim's, i think dong was just saying that prim's doesn't loop through possible choices for the first team.

maggie-j-liu commented on solutions/usaco-531.mdx:
```diff
@@ -122,3 +122,85 @@ int main()
+## Prim's Implementation
+
+	
+### Explanation
```
should be `## Explanation`, move the header and explanation above the Implementation header

maggie-j-liu commented on solutions/usaco-531.mdx:
```diff
@@ -122,3 +122,85 @@ int main()
+
+
+1. Find an unvisited node with the greatest weight, and add it to the MST.
+2. Re-evaluate all the edge weights from the MST  to the other unvisited nodes.
```
```suggestion
2. Re-evaluate all the edge weights from the MST to the other unvisited nodes.
```

dongliu0426 reviewed (commented):
looks good, but imo the format would look better if it was something like

```
## Solution 1 - Kruskal

### Explanation

blah blah blah

### Implementation

blah blah blah

## Solution 2 - Prim

### Explanation

blah blah blah

### Implementation

blah blah blah
```

ryanchou-dev reviewed (approved):
probably what @dongliu0426 suggested, but the rest looks good :)

TruaShamu commented:
done.
